# 01. Data Audit & Exploratory Data Analysis
Auditing the raw parallel datasets from Ghana-NLP for Twi and Ewe.

In [1]:
import sys
sys.path.append("..")
from src.preprocessing import download_raw

# Download and load raw datasets
twi_df = download_raw("Ghana-NLP/ENGLISH_TWI_PARALLEL_TEXT", "twi", text_column="label")
ewe_df = download_raw("Ghana-NLP/EWE_ENGLISH_PARALLEL_TEXT", "ewe", text_column="label")

Repo card metadata block was not found. Setting CardData to empty.


Saved 6090 raw rows for twi to data\raw\twi\raw_sentences.csv


Repo card metadata block was not found. Setting CardData to empty.


Saved 3488 raw rows for ewe to data\raw\ewe\raw_sentences.csv


In [2]:
for name, df in [("Twi", twi_df), ("Ewe", ewe_df)]:
    print(f"\n--- {name} ---")
    print("Columns:", list(df.columns))
    print("Rows:", len(df))
    print("Missing values per column:\n", df.isna().sum())
    print("Duplicate rows:", df.duplicated().sum())
    print("Sample rows:")
    print(df.sample(5, random_state=42))


--- Twi ---
Columns: ['id', 'text', 'label', 'Unnamed: 3']
Rows: 6090
Missing values per column:
 id               0
text             0
label            3
Unnamed: 3    6087
dtype: int64
Duplicate rows: 0
Sample rows:
         id                                               text  \
6049  71084  An education ordinance will be drafted to hand...   
1612  64925  The government is working hand in hand with or...   
1718  70276  Chefs must put on aprons while cooking to get ...   
561   97473  The students were detained because they got in...   
5359  64667  The district will respond to challenges faced ...   

                                                  label Unnamed: 3  
6049  Wɔbɛtwerɛ nhomasua ho mmara bi de adi ɔhaw no ...        NaN  
1612  Aban no ne nnwumakuo ahodoɔ binom redi nsawɔso...        NaN  
1718  Ɛsɛ sɛ aduannoafo hyɛ apron bere a wɔrenoa adu...        NaN  
561   Wɔkyee asukuufoɔ no guu afiaseɛ ɛfiri sɛ na wɔ...        NaN  
5359  Mansini no no bɛdi ɔhaw ahodoɔ ne

### Data Audit Observations & Findings:
1. **Column Mapping & Artifacts**: In both datasets, the English text is stored in `text` and the Ghanaian target language is stored in `label`. The Twi dataset contains an artifact column `Unnamed: 3` with almost all nulls (6,087 / 6,090).
2. **Missing & Incomplete Rows**: There are missing values in the target translation `label` column (3 missing in Twi, 27 missing in Ewe) which must be dropped during preprocessing.
3. **Orthography & Special Characters**: Twi contains unique open vowels (`ɛ`, `ɔ`), while Ewe features rich phonetic and tone orthography (`ɖ`, `ƒ`, `ŋ`, `ɔ`, `ɛ`, `ʋ`, `ɣ`, and diacritics like `lɔlɔ̃tɔa`). Average sentence lengths are typical for parallel translation tasks (~10-25 words per sentence).

## No addition